lib/connection.cpp

```diff
@@ -1771,3 +1890,195 @@ QVector<Connection::SupportedRoomVersion> Connection::availableRoomVersions() co
+    if (!outFile.open(QFile::WriteOnly)) {
+        qCWarning(E2EE) << "Error opening" << outFile.fileName() << ":"
+                        << outFile.errorString();
+        qCWarning(E2EE) << "Caching the rooms state disabled";
+        cacheState = false;
+        return;
+    }
+
+    QJsonObject rootObj {
+        { QStringLiteral("cache_version"),
+          QJsonObject {
+              { QStringLiteral("major"), SyncData::cacheVersion().first },
+              { QStringLiteral("minor"), SyncData::cacheVersion().second } } }
+    };
+    {
+        QJsonArray trackedUsersJson;
+        QJsonArray outdatedUsersJson;
+        for (const auto &user : trackedUsers) {
+            trackedUsersJson += user;
+        }
+        for (const auto &user : outdatedUsers) {
+            outdatedUsersJson += user;
+        }
+        rootObj.insert(QStringLiteral("tracked_users"), trackedUsersJson);
+        rootObj.insert(QStringLiteral("outdated_users"), outdatedUsersJson);
+        QJsonObject devicesList = toJson<QHash<QString, QHash<QString, QueryKeysJob::DeviceInformation>>>(deviceKeys);
+        rootObj.insert(QStringLiteral("devices_list"), devicesList);
+        rootObj.insert(QStringLiteral("sync_token"), q->nextBatchToken());
+    }
+
+#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
+        const auto data =
+            cacheToBinary
+                ? QCborValue::fromJsonValue(rootObj).toCbor()
+                : QJsonDocument(rootObj).toJson(QJsonDocument::Compact);
+#else
+        QJsonDocument json { rootObj };
+        const auto data = cacheToBinary ? json.toBinaryData()
+                                           : json.toJson(QJsonDocument::Compact);
+#endif
+    qCDebug(PROFILER) << "DeviceList generated in" << et;
+
+    outFile.write(data.data(), data.size());
+    qCDebug(E2EE) << "DevicesList saved to" << outFile.fileName();
+}
+
+void Connection::Private::loadDevicesList()
+{
+    QFile file { q->e2eeDataDir() % "/deviceslist.json" };
+    if(!file.exists() || !file.open(QIODevice::ReadOnly)) {
+        qCDebug(E2EE) << "No devicesList cache exists. Creating new";
+        return;
+    }
+    auto data = file.readAll();
+    const auto json = data.startsWith('{')
+                          ? QJsonDocument::fromJson(data).object()
+#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
+                          : QCborValue::fromCbor(data).toJsonValue().toObject()
+#else
+                          : QJsonDocument::fromBinaryData(data).object()
+#endif
+        ;
+    if (json.isEmpty()) {
+        qCWarning(MAIN) << "DevicesList cache is broken or empty, discarding";
+        return;
+    }
+    for(const auto &user : json["tracked_users"].toArray()) {
+        trackedUsers += user.toString();
+    }
+    for(const auto &user : json["outdated_users"].toArray()) {
+        outdatedUsers += user.toString();
+    }
+
+    deviceKeys = fromJson<QHash<QString, QHash<QString, QueryKeysJob::DeviceInformation>>>(json["devices_list"].toObject());
```

Comment:
I have provided a 2-arg `fromJson` overload exactly to avoid such clumsy code. There's a separate `fillFromJson` function template that only fills up parts of the target structure that are provided in the JSON payload.
```suggestion
    fromJson(json["devices_list"], deviceKeys);
```